File: APP/lib/models/pit_scouting_2026.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:scouting_app/models/scout_info.dart';
part 'pit_scouting_2026.freezed.dart';
part 'pit_scouting_2026.g.dart';

@freezed
class PitScouting2026 with _$PitScouting2026 {
  const factory PitScouting2026({
    required String user_id,
    required ScoutInfo scout_info,
    required int team_number,
    required String event_code,
    required double time,
    required PitData2026 data,
    Auto2026? auto,
  }) = _PitScouting2026;
  factory PitScouting2026.fromJson(Map<String, dynamic> json) =>
      _$PitScouting2026FromJson(json);
}

@freezed
class PitData2026 with _$PitData2026 {
  const factory PitData2026(
      {required int driver_experience_events,
      required String drive_train,
      required String type_of_shooter,
      required bool fixedShooting,
      required bool nearTower,
      required bool nearHub,
      required bool go_under_trench,
      required bool can_climb,
      required List<int> climbing,
      required bool can_climb_in_autonomous,
      required String main_strategy,
      required int spare_parts,
      required String favorite_color,
      Auto2026? auto,
      required int hopper_capacity,
      required double bps,
      required String comments,
      // keep autos dynamic to avoid type-mismatch with other code/widgets
      List<Auto2026>? autos,
      required double robot_height,
      required bool straddling_pole_climb_right,
      required bool straddling_pole_climb_left,
      required bool left_pole_climb,
      required bool right_pole_climb,
      required bool center_pole_climb}) = _PitData2026;
  factory PitData2026.fromJson(Map<String, dynamic> json) =>
      _$PitData2026FromJson(json);
}

@freezed
class Auto2026 with _$Auto2026 {
  factory Auto2026({
    required double starting_position_meters_from_hub_center,
    required List<String> field_side,
    required List<AutoStep2026> steps,
    required bool preload,
    required bool climb,
    required bool contacts_robot,
    @Default(false) bool both_sides,
    @JsonKey(name: 'auto_pieces')
    @Default(0)
    int autoPieces, // <-- NEW COUNTER FIELD
  }) = _Auto2026;

  factory Auto2026.fromJson(Map<String, dynamic> json) =>
      _$Auto2026FromJson(json);
}

@freezed
class AutoStep2026 with _$AutoStep2026 {
  const factory AutoStep2026({
    required String name,
    required Map<String, dynamic> extra_data,
  }) = _AutoStep2026;
  factory AutoStep2026.fromJson(Map<String, dynamic> json) =>
      _$AutoStep2026FromJson(json);
}

@freezed
class Data with _$Data {
  factory Data({
    required Auto2026 auto,
  }) = _Data;

  factory Data.fromJson(Map<String, dynamic> json) => _$DataFromJson(json);
}

File: API/models/pit_scouting_2026.py
from pydantic import BaseModel, Field
from typing import List, Optional, Union


from models.scout_info import ScoutInfo



class ExtraData2026(BaseModel):
    position: Union[str, int, None] = None
    shots_from_x: Optional[float] = None
    shots_from_y: Optional[float] = None


class PitAutoStep2026(BaseModel):
    name: str
    extra_data: ExtraData2026
    
class Scoring2026(BaseModel):
    shoot_amount: int = 0
    cycles_completed: int = 0
    shoots_from_X: float = 0
    shoots_from_Y: float = 0
    # Optional fields with defaults for backward compatibility
    feed_amount: Optional[int] = 0
    intake_amount: Optional[int] = 0
    goes_under_trench: Optional[int] = 0
    goes_over_bump: Optional[int] = 0
    climb_side: Optional[str] = "Does not climb"
    # Location-specific scoring fields (for charts)
    # l_1: Optional[int] = 0
    # l_2: Optional[int] = 0
    # l_3: Optional[int] = 0
    # l_4: Optional[int] = 0
    # net: Optional[int] = 0
    # processor: Optional[int] = 0

class Auto2026(BaseModel):
    starting_position_meters_from_hub_center: float
    field_side: List[str]
    preload: bool
    both_sides: bool = False
    contacts_robot: bool = False
    auto_pieces: int = 0
    climb: bool
    steps: List[PitAutoStep2026]

class PitData2026(BaseModel):
    driver_experience_events: int
    drive_train: str
    type_of_shooter: str
    fixedShooting: bool
    nearTower: bool
    nearHub: bool
    go_under_trench: bool
    can_climb: bool
    climbing: list[int]
    can_climb_in_autonomous: bool
    main_strategy: str
    spare_parts: int
    auto_scoring: Scoring2026 = Field(default_factory=Scoring2026)
    favorite_color: str
    autos: list
    hopper_capacity: int

    bps: Optional[float] = None

    robot_height: float

    comments: Optional[str] = None



    straddling_pole_climb_right: bool
    straddling_pole_climb_left: bool
    left_pole_climb: bool
    right_pole_climb: bool
    center_pole_climb: bool



class PitScouting2026(BaseModel):
    user_id: Optional[str] = None
    scout_info: Optional[ScoutInfo] = None
    team_number: Optional[int] = None
    time: Optional[Union[int, float]] = None
    event_code: Optional[str] = None
    data: Optional[PitData2026] = None
